src/ptvsd/adapter/components.py

```diff
@@ -0,0 +1,166 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root
+# for license information.
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+import functools
+
+from ptvsd.common import fmt, json, log, messaging, util
+
+ACCEPT_CONNECTIONS_TIMEOUT = 10
+
+
+class ComponentNotAvailable(Exception):
+    def __init__(self, type):
+        super(ComponentNotAvailable, self).__init__(
+            fmt("{0} is not available", type.__name__)
+        )
+
+
+class Component(util.Observable):
+    """A component managed by a debug adapter: IDE, launcher, or debug server.
+
+    Every component belongs to a Session, which is used for synchronization and
+    shared data.
+
+    Every component has its own message channel, and provides message handlers for
+    that channel. All handlers should be decorated with @Component.message_handler,
+    which ensures that Session is locked for the duration of the handler. Thus, only
+    one handler is running at any given time across all components, unless the lock
+    is released explicitly or via Session.wait_for().
+
+    Components report changes to their attributes to Session, allowing one component
+    to wait_for() a change caused by another component.
+    """
+
+    def __init__(self, session, stream):
+        super(Component, self).__init__()
+
+        self.session = session
+        stream.name = str(self)
+        self.channel = messaging.JsonMessageChannel(stream, self)
+        self.is_connected = True
+
+        self.observers += [lambda *_: session.notify_changed()]
+        self.channel.start()
+
+    def __str__(self):
+        return fmt("{0}-{1}", type(self).__name__, self.session.id)
+
+    @property
+    def ide(self):
+        return self.session.ide
+
+    @property
+    def launcher(self):
+        return self.session.launcher
+
+    @property
+    def server(self):
+        return self.session.server
+
+    def wait_for(self, *args, **kwargs):
+        return self.session.wait_for(*args, **kwargs)
+
+    @staticmethod
+    def message_handler(f):
+        """Applied to a message handler to automatically lock and unlock the session
+        for its duration, and to validate the session state.
+
+        If the handler raises ComponentNotAvailable or JsonIOError, converts it to
+        Message.cant_handle().
+        """
+
+        @functools.wraps(f)
+        def lock_and_handle(self, message):
+            try:
+                with self.session:
+                    return f(self, message)
+            except ComponentNotAvailable as exc:
+                raise message.cant_handle("{0}", exc, silent=True)
+            except messaging.MessageHandlingError as exc:
+                exc.propagate(message)
```

Comment:
Should this be `cant_handle` also?

Reply:
Not necessarily, because it can also be `InvalidMessageError` (it's a subclass) - `exc.propagate()` will take care of all that.